### ISSUE TITLE
Requirements for multi-scheduler to graduate to Beta and then v1

### ISSUE THREAD

**Issue by davidopp** [kubernetes/kubernetes#25318]
We'd like to graduate multi-scheduler to Beta in 1.4.

Beta
- [ ] Change annotation key from `scheduler.alpha.kubernetes.io/name` to `scheduler.beta.kubernetes.io/name`
- [ ] Make sure we have a rollback story or are sure we won't need to rollback earlier than when feature was first introduced

v1
- [ ] Change annotation key from `scheduler.beta.kubernetes.io/name` to `scheduler.kubernetes.io/name`

The design doc has some suggestions for "next steps" and we should incorporate any of those that we actually want to do into "Beta," but I'm not sure any of them are necessary?

ETA: Previously we had "Implement [MetadataPolicy](https://github.com/kubernetes/kubernetes/blob/master/docs/design/metadata-policy.md)" for Beta and "Use initializer pattern in MetadataPolicy" but I think we can separate that from this feature.

cc/ @HaiyangDING @bgrant0607 

ref/ #17865 #18262

**Comment by HaiyangDING** [comment]
I would propose to implement the [multi-policy feature](https://github.com/kubernetes/kubernetes/issues/9920) (which is mentioned in #18262) in parallel and hopefully make it as part of v1.4. The relevant code is almost done already. 

Also, I am glad to implement the MetadataPolicy (API & Admission controller) feature (already started). Just one question, when is the v1.4 going to be released?

**Comment by AdoHe** [comment]
Also if possible, I would take a look at MetadataPolicy.

**Comment by therc** [comment]
Would it make sense to also make pods belong to a scheduler given the resources it requests, without requiring an annotation? I'm thinking in particular of cases like GPUs or custom resource types. If I add a scheduler that declares being able to handle GPU resources and e.g. custom resources (through a regex?), annotations could be redundant. This wouldn't apply to all instances where there are multiple schedulers, of course, but would simplify certain kinds of deployments.

By a brief look at MetadataPolicy, perhaps this could be an additional admission controller? What I've seen so far about multiple schedulers seems to be focused on QoS scenarios, but it would be nice to take into consideration us folks with esoteric resources as well. :-)

**Comment by davidopp** [comment]
Yeah, I'm a little worried that MetadataPolicy is too restrictive, but anyway, as you say, you could write an admission controller that adds the scheduler name annotation based on the presence or absence of a particular resource (this would work particularly well with #19082). We plan to make it possible to add admission controllers dynamically (without recompiling the system) but it hasn't bubbled to the top of the priority list yet.

**Comment by gmarek** [comment]
Moved out of 1.5, so I'm unassigning myself.

**Comment by dalanlan** [comment]
Any update on this thread?

**Comment by davidopp** [comment]
In general the "move XXX to beta" features where XXX is currently expressed using an annotation, are waiting for a decision on how we want field-level versioning to work (kubernetes/kubernetes#34508). If we can find someone to do multi-scheduler to beta for 1.6 then we will, but affinity and taints/tolerations are higher on the priority list based on the feedback collected from sig-scheduling.

**Comment by timothysc** [comment]
I'm uncertain if we want to stretch this into 1.6 unless @ConnorDoyle || @balajismaniam have cycles here.

**Comment by k82cn** [comment]
@timothysc , I have cycles for this issues, but need a shepherd :).

**Comment by timothysc** [comment]
I can shepherd it in.  You can follow the pattern alpha annotations -> beta fields as seen in : https://github.com/kubernetes/kubernetes/pull/37299

**Comment by k82cn** [comment]
Thanks very much; I'm going to work on that.

**Comment by davidopp** [comment]
Let's leave this issue open since it is about "beta and v1" (GA)

Thanks for your work @k82cn !

**Comment by davidopp** [comment]
AFAIK there is no remaining work for moving multi-scheduler to Beta (though we really could use some better testing). #38871 by @k82cn replaced the annotation with a field.

Upgrading to 1.6 will break existing users of this feature and they will have to switch to fields, but this within the bounds of our policy for alpha features. Information about upgrading has been sent to kubernetes-dev ([here](https://groups.google.com/d/msg/kubernetes-dev/-fwfHQtm9GE/H1TU4yqiBAAJ)) and kubernetes-users ([here](https://groups.google.com/d/msg/kubernetes-users/xf-VlRlAm2A/Th76aF-lAwAJ)) and will also appear in the release notes.

**Comment by enlinxu** [comment]
Does it make sense to expose an argument to kubectl create , to allow a more dynamic way to decide the scheduler?

**Comment by timothysc** [comment]
@enlinxu If there is enough demand to plumb through to command line, sure. 

Could you describe your use case on why you wouldn't just adjust your pod-spec to denote?

**Comment by enlinxu** [comment]
@timothysc The use case is to allow user to have a more dynamic and runtime way to start a pod with scheduler foo on the fly, without checking the pod-spec every time.
